OpenRA.Mods.Common/Traits/PowerTooltip.cs

```diff
@@ -0,0 +1,50 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Shown power info on the build palette widget.")]
+	public class PowerTooltipInfo : UpgradableTraitInfo
+	{
+		public override object Create(ActorInitializer init) { return new PowerTooltip(init.Self); }
+	}
+
+	public class PowerTooltip : IProvideTooltipInfo, INotifyOwnerChanged
+	{
+		readonly Actor self;
+		PowerManager powerManager;
+
+		public PowerTooltip(Actor self)
+		{
+			this.self = self;
+			powerManager = self.Owner.PlayerActor.Trait<PowerManager>();
+		}
+
+		public bool IsTooltipVisible(Player forPlayer)
+		{
+			return forPlayer == self.Owner;
+		}
+
+		public string TooltipText
+		{
+			get
+			{
+				return "Power Usage: {0}{1}".F(powerManager.PowerDrained, powerManager.PowerProvided != 1000000 ? "/" + powerManager.PowerProvided : "");
```

Comment:
It is/was a magic value for the unlimited power cheat. It should be replaced with something like `self.Owner.PlayerActor.Trait<DeveloperMode>().UnlimitedPower`. Actually, `self.Owner.PlayerActor.Trait<DeveloperMode>()` should be cached. See OpenRA.Mods.Common/Traits/Power/Player/PowerManager.cs